File: android/src/main/kotlin/io/github/kokemus/cross_bluetooth_api/CrossBluetoothApiPlugin.kt
package io.github.kokemus.cross_bluetooth_api

import android.Manifest
import android.bluetooth.BluetoothGattCharacteristic
import android.os.Handler
import android.os.Looper
import androidx.annotation.RequiresPermission
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.github.kokemus.cross_bluetooth_api.commands.*
import io.github.kokemus.cross_bluetooth_api.events.CharacteristicValueChangedEvent
import io.github.kokemus.cross_bluetooth_api.events.GattServerDisconnectedEvent
import io.github.kokemus.cross_bluetooth_api.services.BluetoothManagerImp
import io.github.kokemus.cross_bluetooth_api.services.BluetoothManager
import io.github.kokemus.cross_bluetooth_api.services.BluetoothManagerListener
import io.github.kokemus.cross_bluetooth_api.services.DeviceManager
import io.github.kokemus.cross_bluetooth_api.services.DeviceManagerListener
import io.github.kokemus.cross_bluetooth_api.services.RequestDeviceLauncher
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch

class CrossBluetoothApiPlugin : FlutterPlugin, MethodCallHandler, ActivityAware, BluetoothManagerListener,
  DeviceManagerListener {
  private lateinit var channel : MethodChannel
  private lateinit var bluetoothManager: BluetoothManager
  private val requestDeviceLauncher = RequestDeviceLauncher()
  private val commandScope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)

  override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    channel = MethodChannel(flutterPluginBinding.binaryMessenger, "cross_bluetooth_api")
    channel.setMethodCallHandler(this)

    bluetoothManager = BluetoothManagerImp(flutterPluginBinding.applicationContext)
    bluetoothManager.addListener(this)

    val eventChannel = EventChannel(flutterPluginBinding.binaryMessenger, "cross_bluetooth_api/events")
    eventChannel.setStreamHandler(streamHandler)
  }

  internal var eventSink: EventChannel.EventSink? = null
  internal val handler = Handler(Looper.getMainLooper())

  private val streamHandler = object: EventChannel.StreamHandler {
    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
      eventSink = events
    }

    override fun onCancel(arguments: Any?) {
      eventSink = null
    }
  }

  override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    bluetoothManager.removeListener(this)
    bluetoothManager.removeAllListeners()
    commandScope.cancel()
    channel.setMethodCallHandler(null)
  }

  override fun onAttachedToActivity(activityPluginBinding: ActivityPluginBinding) {
    requestDeviceLauncher.attach(activityPluginBinding)
  }

  override fun onDetachedFromActivity() {
    requestDeviceLauncher.detach()
  }

  override fun onReattachedToActivityForConfigChanges(activityPluginBinding: ActivityPluginBinding) {
    requestDeviceLauncher.attach(activityPluginBinding)
  }

  override fun onDetachedFromActivityForConfigChanges() {
    requestDeviceLauncher.detach()
  }

  @RequiresPermission(Manifest.permission.BLUETOOTH_CONNECT)
  override fun onMethodCall(call: MethodCall, result: Result) {
    @Suppress("UNCHECKED_CAST")
    val arguments = call.arguments as? Map<String, Any> ?: emptyMap()

    val command = when (CommandId.fromMethod(call.method)) {
      CommandId.REQUEST_DEVICE -> RequestDeviceCommand(requestDeviceLauncher, arguments, result)
      CommandId.CONNECT -> ConnectCommand(bluetoothManager, arguments, result)
      CommandId.DISCONNECT -> DisconnectCommand(bluetoothManager, arguments, result)
      CommandId.GET_PRIMARY_SERVICE -> GetPrimaryServiceCommand(bluetoothManager, arguments, result)
      CommandId.GET_CHARACTERISTIC -> GetCharacteristicCommand(bluetoothManager, arguments, result)
      CommandId.READ_VALUE -> ReadValueCommand(bluetoothManager, arguments, result)
      CommandId.WRITE_VALUE_WITHOUT_RESPONSE -> WriteValueWithoutResponseCommand(bluetoothManager, arguments, result)
      CommandId.WRITE_VALUE_WITH_RESPONSE -> WriteValueWithResponseCommand(bluetoothManager, arguments, result)
      CommandId.START_NOTIFICATIONS -> StartNotificationsCommand(bluetoothManager, arguments, result)
      CommandId.STOP_NOTIFICATIONS -> StopNotificationsCommand(bluetoothManager, arguments, result)
      null -> null
    }

    if (command == null) {
      result.notImplemented()
      return
    }

    commandScope.launch {
      command.execute()
    }
  }

  override fun onConnected(deviceManager: DeviceManager) {
    deviceManager.addListener(this)
  }

  override fun onDisconnected(deviceId: String) {
    handler.post {
      eventSink?.success(GattServerDisconnectedEvent(deviceId).toMap())
    }
  }

  override fun onConnectFailed(deviceId: String) {}

  override fun onCharacteristicChanged(
    deviceManager: DeviceManager,
    characteristic: BluetoothGattCharacteristic,
    value: ByteArray
  ) {
    handler.post {
      eventSink?.success(
        CharacteristicValueChangedEvent(
          deviceId = deviceManager.deviceId,
          serviceUUID = characteristic.service.uuid.toString(),
          characteristicUUID = characteristic.uuid.toString(),
          value = value
        ).toMap()
      )
    }
  }
}

File: ios/cross_bluetooth_api/Sources/cross_bluetooth_api/CrossBluetoothApiPlugin.swift
import Flutter
import UIKit
import CoreBluetooth

public class SwiftCrossBluetoothApiPlugin: NSObject, FlutterPlugin {
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "cross_bluetooth_api", binaryMessenger: registrar.messenger())
        
        let eventChannel = FlutterEventChannel(name: "cross_bluetooth_api/events", binaryMessenger: registrar.messenger())

        let instance = SwiftCrossBluetoothApiPlugin()
        eventChannel.setStreamHandler(instance)
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    private var viewController: UIViewController {
        let windowScenes = UIApplication.shared.connectedScenes
            .compactMap { $0 as? UIWindowScene }
        let keyWindow = windowScenes
            .flatMap { $0.windows }
            .first { $0.isKeyWindow }

        return keyWindow!.rootViewController!
    }

    private var eventSink: FlutterEventSink?
    private var manager: BluetoothManager!

    override init() {
        super.init()
        manager = BluetoothManagerImp(queue: .main)
        manager.addDelegate(self)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let method = CommandId(rawValue: call.method)
        let arguments = call.arguments as? [String: AnyObject]
        let command: Command? = switch method {
        case .requestDevice:
            RequestDeviceCommand(
                viewController: viewController,
                arguments: arguments,
                pendingResult: result
            )
        case .connect:
            ConnectCommand(
                manager: manager,
                arguments: arguments,
                pendingResult: result
            )
        case .disconnect:
            DisconnectCommand(
                manager: manager,
                arguments: arguments,
                pendingResult: result
            )
        case .getPrimaryService:
            GetPrimaryServiceCommand(
                manager: manager,
                arguments: arguments,
                pendingResult: result
            )
        case .getCharacteristic:
            GetCharacteristicCommand(
                manager: manager,
                arguments: arguments,
                pendingResult: result
            )
        case .readValue:
            ReadValueCommand(
                manager: manager,
                arguments: arguments,
                pendingResult: result
            )
        case .writeValueWithoutResponse:
            WriteValueWithoutResponseCommand(
                manager: manager,
                arguments: arguments,
                pendingResult: result
            )
        case .writeValueWithResponse:
            WriteValueWithResponseCommand(
                manager: manager,
                arguments: arguments,
                pendingResult: result
            )
        case .startNotifications:
            StartNotificationsCommand(
                manager: manager,
                arguments: arguments,
                pendingResult: result
            )
        case .stopNotifications:
            StopNotificationsCommand(
                manager: manager,
                arguments: arguments,
                pendingResult: result
            )
        default:
            nil
        }

        if let command {
            Task {
                await command.execute()
            }
        } else {
            result(FlutterMethodNotImplemented)
        }
    }
}

extension SwiftCrossBluetoothApiPlugin: FlutterStreamHandler {
    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        eventSink = events
        return nil
    }
    
    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        eventSink = nil
        return nil
    }
}

extension SwiftCrossBluetoothApiPlugin: BluetoothManagerDelegete {
    public func centralManagerDidUpdateState(_ central: CBCentralManager) {}

    public func centralManager(_ central: CBCentralManager, didConnect peripheral: CBPeripheral) {
        manager.addPeripheral(peripheral).addDelegate(self)
    }

    public func centralManager(_ central: CBCentralManager, didFailToConnect peripheral: CBPeripheral, error: Error?) {}

    public func centralManager(_ central: CBCentralManager, didDisconnectPeripheral peripheral: CBPeripheral, error: Error?) {
        manager.removePeripheral(peripheral)
        eventSink?(
            GattServerDisconnectedEvent(
                deviceId: peripheral.identifier.uuidString
            ).toMap()
        )
    }
}

extension SwiftCrossBluetoothApiPlugin: DeviceManagerDelegate {
    public func peripheral(_ peripheral: CBPeripheral, didDiscoverServices error: Error?) {}

    public func peripheral(_ peripheral: CBPeripheral, didDiscoverCharacteristicsFor service: CBService, error: Error?) {}

    public func peripheral(_ peripheral: CBPeripheral, didUpdateValueFor characteristic: CBCharacteristic, error: Error?) {
        if error == nil {
            eventSink?(
                CharacteristicValueChangedEvent(
                    deviceId: peripheral.identifier.uuidString,
                    serviceUUID: characteristic.service!.uuid.uuidString,
                    characteristicUUID: characteristic.uuid.uuidString,
                    value: characteristic.value ?? Data()
                ).toMap()
            )
        }
    }

    public func peripheral(_ peripheral: CBPeripheral, didWriteValueFor characteristic: CBCharacteristic, error: Error?) {}
}
